File: src/Roguelike/Core/Combat/ActionResolver.cs
using Roguelike.Data;
using System;
using System.Linq;

namespace Roguelike.Core
{
    /// <summary>
    /// A static utility class responsible for calculating and applying combat actions.
    /// It handles the math for Damage (Strength, Vulnerable) and Block (Frail),
    /// as well as the application of status and deck effects.
    /// </summary>
    public static class ActionResolver
    {
        /// <summary>
        /// Resolves a single action against a specific target.
        /// </summary>
        /// <param name="action">The action data to execute.</param>
        /// <param name="source">The combatant performing the action.</param>
        /// <param name="target">The combatant receiving the action.</param>
        /// <param name="getEffectById">A function/delegate to look up EffectData by its string ID.</param>
        public static void Resolve(CombatActionData action, Combatant source, Combatant target, Func<string, EffectData> getEffectById)
        {
            switch (action.Type)
            {
                case ActionType.DealDamage:
                    ApplyDamage(action.Value, source, target);
                    break;

                case ActionType.GainBlock:
                    ApplyBlock(action.Value, source, target);
                    break;

                case ActionType.ApplyStatusEffect:
                    ApplyStatusEffect(action.EffectId, target, getEffectById);
                    break;

                case ActionType.ApplyDeckEffect:
                    ApplyDeckEffect(action.EffectId, action.Value, target, getEffectById);
                    break;
            }
        }

        private static void ApplyDamage(int baseDamage, Combatant source, Combatant target)
        {
            float finalDamage = baseDamage;

            // Apply all Strength effects
            foreach (var effect in source.ActiveEffects)
            {
                if (effect.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Strength)
                {
                    if (s.IntensityType == IntensityType.Flat)
                    {
                        finalDamage += s.Intensity;
                    }
                    else // Percentage
                    {
                        finalDamage *= (1 + s.Intensity / 100f);
                    }
                }
            }

            // Check for Weakened (reduces outgoing damage)
            var weakEffect = source.ActiveEffects.FirstOrDefault(e => 
                e.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Weakened);
            
            if (weakEffect != null)
            {
                var weakData = (StatusEffectData)weakEffect.SourceData;
                if (weakData.IntensityType == IntensityType.Percentage)
                {
                    // Intensity represents reduction percentage (e.g., 25 = -25% damage = 0.75x multiplier)
                    finalDamage *= (1 - weakData.Intensity / 100f);
                }
            }

            // Check for Vulnerable (increases incoming damage)
            var vulEffect = target.ActiveEffects.FirstOrDefault(e => 
                e.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Vulnerable);
                
            if (vulEffect != null)
            {
                var vulData = (StatusEffectData)vulEffect.SourceData;
                if (vulData.IntensityType == IntensityType.Percentage)
                {
                    // Intensity represents increase percentage (e.g., 50 = +50% damage = 1.5x multiplier)
                    finalDamage *= (1 + vulData.Intensity / 100f);
                }
            }

            int damageInt = (int)Math.Floor(finalDamage);
            if (damageInt < 0) damageInt = 0;

            // Check if target has Pierced effect
            bool hasPierced = target.ActiveEffects.Any(e => 
                e.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Pierced);
            
            if (hasPierced)
            {
                target.TakePiercingDamage(damageInt);
            }
            else
            {
                target.TakeDamage(damageInt);
            }
        }

        private static void ApplyBlock(int baseBlock, Combatant source, Combatant target)
        {
            float finalBlock = baseBlock;

            // Check for Frail (reduces block gained)
            var frailEffect = target.ActiveEffects.FirstOrDefault(e => 
                e.SourceData is StatusEffectData s && s.EffectType == StatusEffectType.Frail);

            if (frailEffect != null)
            {
                var frailData = (StatusEffectData)frailEffect.SourceData;
                if (frailData.IntensityType == IntensityType.Percentage)
                {
                    // Intensity represents reduction percentage (e.g., 25 = -25% block = 0.75x multiplier)
                    finalBlock *= (1 - frailData.Intensity / 100f);
                }
            }

            int blockInt = (int)Math.Floor(finalBlock);
            target.GainBlock(blockInt);
        }

        private static void ApplyStatusEffect(string effectId, Combatant target, Func<string, EffectData> getEffectById)
        {
            var effectData = getEffectById(effectId);
            if (effectData != null && effectData is StatusEffectData)
            {
                target.ApplyEffect(effectData);
            }
        }

        private static void ApplyDeckEffect(string effectId, int value, Combatant target, Func<string, EffectData> getEffectById)
        {
            if (target is Hero hero)
            {
                var effectData = getEffectById(effectId);
                if (effectData != null && effectData is DeckEffectData deckEffect)
                {
                    switch (deckEffect.EffectType)
                    {
                        case DeckEffectType.DrawCard:
                            hero.Deck.DrawCards(value);
                            break;
                        case DeckEffectType.DiscardCard:
                            if (hero.Deck.Hand.Count > 0)
                            {
                                var card = hero.Deck.Hand[0]; 
                                hero.Deck.DiscardCardFromHand(card);
                            }
                            break;
                        case DeckEffectType.FreezeCard:
                            // TODO: Implement card freezing logic
                            break;
                        case DeckEffectType.DuplicateCard:
                            // TODO: Implement card duplication logic
                            break;
                    }
                }
            }
        }
    }
}

File: python_roguelike/core/action_resolver.py
import math
from typing import Callable, Optional
from .combatant import Combatant
from .active_effect import ActiveEffect
from ..data.combat_action_data import CombatActionData
from ..data.effect_data import EffectData
from ..data.status_effect_data import StatusEffectData
from ..data.deck_effect_data import DeckEffectData
from ..data.enums import (
    ActionType, StatusEffectType, DeckEffectType, IntensityType, DecayType
)


class ActionResolver:

    @staticmethod
    def resolve(
        action: CombatActionData,
        source: Combatant,
        target: Combatant,
        get_effect_by_id: Callable[[str], Optional[EffectData]]
    ):
        if action.type == ActionType.DealDamage:
            ActionResolver._apply_damage(action.value, source, target)
        elif action.type == ActionType.GainBlock:
            ActionResolver._apply_block(action.value, source, target)
        elif action.type == ActionType.GainHealth:
            ActionResolver._apply_heal(action.value, target)
        elif action.type == ActionType.ApplyStatusEffect:
            ActionResolver._apply_status_effect(action.effect_id, target, get_effect_by_id)
        elif action.type == ActionType.ApplyDeckEffect:
            ActionResolver._apply_deck_effect(action.effect_id, action.value, target, get_effect_by_id)

    @staticmethod
    def _apply_damage(base_damage: int, source: Combatant, target: Combatant):
        final_damage = float(base_damage)

        # Apply Strength
        for ae in source.active_effects:
            if isinstance(ae.source_data, StatusEffectData) and ae.source_data.effect_type == StatusEffectType.Strength:
                if ae.source_data.intensity_type == IntensityType.Flat:
                    final_damage += ae.source_data.intensity
                else:
                    final_damage *= (1 + ae.source_data.intensity / 100.0)

        # Apply Weakened (source)
        weak_ae = next(
            (ae for ae in source.active_effects
             if isinstance(ae.source_data, StatusEffectData) and ae.source_data.effect_type == StatusEffectType.Weakened),
            None
        )
        if weak_ae is not None:
            wd = weak_ae.source_data
            if wd.intensity_type == IntensityType.Percentage:
                final_damage *= (1 - wd.intensity / 100.0)

        # Apply Vulnerable (target)
        vul_ae = next(
            (ae for ae in target.active_effects
             if isinstance(ae.source_data, StatusEffectData) and ae.source_data.effect_type == StatusEffectType.Vulnerable),
            None
        )
        if vul_ae is not None:
            vd = vul_ae.source_data
            if vd.intensity_type == IntensityType.Percentage:
                final_damage *= (1 + vd.intensity / 100.0)

        damage_int = int(math.floor(final_damage))
        if damage_int < 0:
            damage_int = 0

        # Check Pierced
        has_pierced = any(
            isinstance(ae.source_data, StatusEffectData) and ae.source_data.effect_type == StatusEffectType.Pierced
            for ae in target.active_effects
        )
        if has_pierced:
            target.take_piercing_damage(damage_int)
        else:
            target.take_damage(damage_int)

    @staticmethod
    def _apply_block(base_block: int, source: Combatant, target: Combatant):
        final_block = float(base_block)

        frail_ae = next(
            (ae for ae in target.active_effects
             if isinstance(ae.source_data, StatusEffectData) and ae.source_data.effect_type == StatusEffectType.Frail),
            None
        )
        if frail_ae is not None:
            fd = frail_ae.source_data
            if fd.intensity_type == IntensityType.Percentage:
                final_block *= (1 - fd.intensity / 100.0)

        block_int = int(math.floor(final_block))
        target.gain_block(block_int)

    @staticmethod
    def _apply_heal(value: int, target: Combatant):
        target.heal(value)

    @staticmethod
    def _apply_status_effect(
        effect_id: str,
        target: Combatant,
        get_effect_by_id: Callable[[str], Optional[EffectData]]
    ):
        effect_data = get_effect_by_id(effect_id)
        if effect_data is not None and isinstance(effect_data, StatusEffectData):
            target.apply_effect(effect_data)

    @staticmethod
    def _apply_deck_effect(
        effect_id: str,
        value: int,
        target: Combatant,
        get_effect_by_id: Callable[[str], Optional[EffectData]]
    ):
        from .hero import Hero
        if not isinstance(target, Hero):
            return

        hero = target
        effect_data = get_effect_by_id(effect_id)
        if effect_data is not None and isinstance(effect_data, DeckEffectData):
            if effect_data.effect_type == DeckEffectType.DrawCard:
                hero.deck.draw_cards(value)
            elif effect_data.effect_type == DeckEffectType.DiscardCard:
                if hero.deck.hand:
                    hero.deck.discard_card_from_hand(hero.deck.hand[0])
            elif effect_data.effect_type == DeckEffectType.FreezeCard:
                pass  # TODO
            elif effect_data.effect_type == DeckEffectType.DuplicateCard:
                pass  # TODO
